pkg/extendapi/devicetemplate.go

```diff
@@ -0,0 +1,213 @@
+		logrus.Error("DeviceTemplateHandler request GetAuthName error:", err.Error())
+		h.WriteResponse(w, err.Error())
+		return
+	}
+
+	var request DeviceTemplateRequest
+
+	if err := json.Unmarshal(body, &request); err != nil {
+		logrus.Error("DeviceTemplateHandler request body validate json error:", err.Error())
+		h.WriteResponse(w, err.Error())
+		return
+	}
+
+	if err := h.validateRequest(request); err != nil {
+		logrus.Error("DeviceTemplateHandler request body validate error:", err.Error())
+		h.WriteResponse(w, err.Error())
+		return
+	}
+
+	if err := h.validTemplate(r.Context(), request); err != nil {
+		logrus.Error("DeviceTemplateHandler validTemplate error:", err.Error())
+		h.WriteResponse(w, err.Error())
+		return
+	}
+
+	if err := h.createSecret(r.Context(), request, user); err != nil {
+		logrus.Error("DeviceTemplateHandler createSecret error:", err.Error())
+		h.WriteResponse(w, err.Error())
+		return
+	}
+
+	h.WriteResponse(w, "success")
+}
+
+func (h *DeviceTemplateHandler) validateRequest(req DeviceTemplateRequest) error {
+
+	if req.DeviceType == "" {
+		return errors.New("request body have no field deviceType")
+	}
+
+	if req.TemplateName == "" {
+		return errors.New("request body have no field templateName")
+	}
+
+	if req.Namespace == "" {
+		return errors.New("request body have no field namespace")
+	}
+
+	if req.Version == "" {
+		return errors.New("request body have no field version")
+	}
+
+	if req.DeviceGroup == "" {
+		return errors.New("request body have no field deviceGroup")
+	}
+
+	if req.ResourceName == "" {
+		return errors.New("request body have no field resourceName")
+	}
+
+	if req.TemplateSpec == nil {
+		return errors.New("request body have no field spec")
+	}
+
+	return nil
+}
+
+func (h *DeviceTemplateHandler) GetAuthName(r *http.Request) (string, error) {
+	tokenAuthValue := auth.GetTokenAuthFromRequest(r)
+	if tokenAuthValue == "" {
+		return "", errors.New("must authenticate")
+	}
+
+	tokenName, tokenKey := auth.SplitTokenParts(tokenAuthValue)
+
+	if tokenName == "" || tokenKey == "" {
+		return "", errors.New("must authenticate")
+	}
+
+	return tokenName, nil
+}
+
+func (h *DeviceTemplateHandler) WriteResponse(w http.ResponseWriter, msg string) {
+	io.WriteString(w, fmt.Sprintf(`{"msg":%s}`, msg))
+}
+
+func (h *DeviceTemplateHandler) createSecret(ctx context.Context, req DeviceTemplateRequest, user string) error {
+
+	js, err := req.TemplateSpec.MarshalJSON()
+	if err != nil {
+		return err
+	}
+
+	stringData := map[string]string{"template": string(js)}
+
+	tempStr := util.GenerateRandomTempKey(7)
+
+	name := fmt.Sprintf("%s-template-%s", strings.ToLower(req.DeviceType), tempStr)
+
+	secret := &apiv1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   req.Namespace,
+			Annotations: map[string]string{"user": user, "deviceType": req.DeviceType, "templateName": req.TemplateName},
```

Comment:
set the annotation key(use constant) to be something meaningful like below to avoid duplicate and use better format
``` 
edgeapi.cattle.io/device-template-name: name
edgeapi.cattle.io/device-template-type:  type
edgeapi.cattle.io/device-template-creator:  user
```